README.md
# `dm_fast_mapping`: DeepMind Fast Language Learning Tasks

The *DeepMind Fast Language Learning Tasks* is a set of machine-learning tasks
that requires agents to learn the meaning of instruction words either slowly
(i.e. across many episodes), quickly (i.e. within a single episode) or both.

The tasks in this repo are [Unity-based](http://unity3d.com/).

## Overview

These tasks are provided through pre-packaged
[Docker containers](http://www.docker.com).

This package consists of support code to run these Docker containers. You
interact with the task environment via a
[`dm_env`](http://www.github.com/deepmind/dm_env) Python interface.

Please see the [documentation](docs/index.md) for more detailed information on
the available tasks, actions and observations.

## Requirements

`dm_fast_mapping` requires [Docker](https://www.docker.com),
[Python](https://www.python.org/) 3.6.1 or later and a x86-64 CPU with SSE4.2
support. We do not attempt to maintain a working version for Python 2.

Note: We recommend using
[Python virtual environment](https://docs.python.org/3/tutorial/venv.html) to
mitigate conflicts with your system's Python environment.

Download and install Docker:

*   For Linux, install [Docker-CE](https://docs.docker.com/install/)
*   Install Docker Desktop for
    [OSX](https://docs.docker.com/docker-for-mac/install/) or
    [Windows](https://docs.docker.com/docker-for-windows/install/).

## Installation

You can install `dm_fast_mapping` by cloning a local copy of our GitHub
repository:

```bash
$ git clone https://github.com/deepmind/dm_fast_mapping.git
$ pip install ./dm_fast_mapping
```

You can install the dependencies for the `examples/` with:

```bash
$ pip install ./dm-fast-mapping[examples]
```

## Usage

Once `dm_fast_mapping` is installed, to instantiate a `dm_env` instance run the
following:

```python
import dm_fast_mapping

settings = dm_fast_mapping.EnvironmentSettings(seed=123,
    level_name='fast_slow/fast_map_three_objs')
env = dm_fast_mapping.load_from_docker(settings)
```

## Citing

If you use `dm_fast_mapping` in your work, please cite the accompanying paper:

```bibtex
@misc{hill2020grounded,
      title={Grounded Language Learning Fast and Slow},
      author={Felix Hill and
              Olivier Tieleman and
              Tamara von Glehn and
              Nathaniel Wong and
              Hamza Merzic and
              Stephen Clark},
      year={2020},
      eprint={2009.01719},
      archivePrefix={arXiv},
      primaryClass={cs.CL}
}
```

For the `with_distractors` tasks, please also cite the source for those tasks:

```bibtex
@misc{lampinen2021towards,
      title={Towards mental time travel:
             a hierarchical memory for reinforcement learning agents},
      author={Lampinen, Andrew Kyle and
              Chan, Stephanie C Y and
              Banino, Andrea and
              Hill, Felix},
      archivePrefix={arXiv},
      eprint={2105.14039},
      year={2021},
      primaryClass={cs.LG}
}
```

## Notice

This is not an officially supported Google product.


RELEASE_NOTES.md
# Release Notes

## [1.0.0]

*   Initial release.

## [1.1.0]

*   Added tasks from the paper "Towards mental time travel: a hierarchical
    memory for reinforcement learning agents."

## [1.1.1]

*   Added additional tasks from the revised paper "Towards mental time travel: a
    hierarchical memory for reinforcement learning agents."


dm_fast_mapping/load_from_disk_test.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Tests for dm_fast_mapping.load_from_disk."""

from absl import flags
from absl.testing import absltest
from absl.testing import parameterized
from dm_env import test_utils
import dm_fast_mapping

FLAGS = flags.FLAGS

flags.DEFINE_string('path', '',
                    'Directory that contains dm_fast_mapping environment.')


class LoadFromDiskTest(test_utils.EnvironmentTestMixin, absltest.TestCase):

  def make_object_under_test(self):
    return dm_fast_mapping.load_from_disk(
        FLAGS.path,
        settings=dm_fast_mapping.EnvironmentSettings(
            seed=123, level_name='architecture_comparison/fast_map_three_objs'))


class FastMappingTaskTest(parameterized.TestCase):

  @parameterized.parameters(dm_fast_mapping.FAST_MAPPING_TASK_LEVEL_NAMES)
  def test_load_level(self, level_name):
    self.assertIsNotNone(
        dm_fast_mapping.load_from_disk(
            FLAGS.path,
            settings=dm_fast_mapping.EnvironmentSettings(
                seed=123, level_name=level_name)))


if __name__ == '__main__':
  absltest.main()


dm_fast_mapping/load_from_docker_test.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Tests for dm_fast_mapping.load_from_docker."""

from absl import flags
from absl.testing import absltest
from absl.testing import parameterized
from dm_env import test_utils
import dm_fast_mapping

FLAGS = flags.FLAGS

flags.DEFINE_string(
    'docker_image_name', None,
    'Name of the Docker image that contains the Fast Language Learning Tasks. '
    'If None, uses the default dm_fast_mapping name')


class LoadFromDockerTest(test_utils.EnvironmentTestMixin, absltest.TestCase):

  def make_object_under_test(self):
    return dm_fast_mapping.load_from_docker(
        name=FLAGS.docker_image_name,
        settings=dm_fast_mapping.EnvironmentSettings(
            seed=123, level_name='architecture_comparison/fast_map_three_objs'))


class FastMappingTaskTest(parameterized.TestCase):

  @parameterized.parameters(dm_fast_mapping.FAST_MAPPING_TASK_LEVEL_NAMES)
  def test_load_level(self, level_name):
    self.assertIsNotNone(
        dm_fast_mapping.load_from_docker(
            name=FLAGS.docker_image_name,
            settings=dm_fast_mapping.EnvironmentSettings(
                seed=123, level_name=level_name)))


if __name__ == '__main__':
  absltest.main()


examples/random_agent.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Example random agent for interacting with DeepMind Fast Mapping Tasks."""

from absl import app
from absl import flags
from absl import logging
from dm_env import specs
import dm_fast_mapping
import numpy as np

FLAGS = flags.FLAGS

flags.DEFINE_string(
    'docker_image_name', None,
    'Name of the Docker image that contains the Fast Language Learning Tasks. '
    'If None, uses the default dm_fast_mapping name')

flags.DEFINE_integer('seed', 123, 'Environment seed.')
flags.DEFINE_string('level_name', 'fast_slow/fast_map_three_objs',
                    'Name of task to run.')


class RandomAgent(object):
  """Basic random agent for DeepMind Fast Language Fast Language Learning Tasks."""

  def __init__(self, action_spec):
    self.action_spec = action_spec

  def act(self):
    action = {}

    for name, spec in self.action_spec.items():
      # Uniformly sample BoundedArray actions.
      if isinstance(spec, specs.BoundedArray):
        action[name] = np.random.uniform(spec.minimum, spec.maximum, spec.shape)
      else:
        action[name] = spec.generate_value()
    return action


def main(_):
  if 'with_distractors' in FLAGS.level_name:  # for the tasks from the HTM paper
    episode_length_seconds = 450.0
  else:
    episode_length_seconds = 120.0

  env_settings = dm_fast_mapping.EnvironmentSettings(
      seed=FLAGS.seed, level_name=FLAGS.level_name,
      episode_length_seconds=episode_length_seconds)
  with dm_fast_mapping.load_from_docker(
      name=FLAGS.docker_image_name, settings=env_settings) as env:
    agent = RandomAgent(env.action_spec())

    timestep = env.reset()
    score = 0
    while not timestep.last():
      action = agent.act()
      timestep = env.step(action)

      if timestep.reward:
        score += timestep.reward
        logging.info('Total score: %1.1f, reward: %1.1f', score,
                     timestep.reward)


if __name__ == '__main__':
  app.run(main)


dm_fast_mapping/__init__.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Python utilities for running dm_fast_mapping."""

from dm_fast_mapping import _load_environment
from dm_fast_mapping._version import __version__

EnvironmentSettings = _load_environment.EnvironmentSettings

FAST_MAPPING_TASK_LEVEL_NAMES = _load_environment.FAST_MAPPING_TASK_LEVEL_NAMES

load_from_disk = _load_environment.load_from_disk
load_from_docker = _load_environment.load_from_docker


dm_fast_mapping/_version.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Package version for dm_fast_mapping.

Kept in separate file so it can be used during installation.
"""

__version__ = '1.0.0'  # https://www.python.org/dev/peps/pep-0440/


dm_fast_mapping/_load_environment.py
# Lint as: python3
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Python utility functions for loading DeepMind Fast Language Learning Tasks."""

import codecs
import collections
import json
import os
import re
import subprocess
import time
import typing

from absl import logging
import dm_env
import docker
import grpc
import numpy as np
import portpicker

from dm_env_rpc.v1 import connection as dm_env_rpc_connection
from dm_env_rpc.v1 import dm_env_adaptor
from dm_env_rpc.v1 import dm_env_rpc_pb2
from dm_env_rpc.v1 import error
from dm_env_rpc.v1 import tensor_utils

# Maximum number of times to attempt gRPC connection.
_MAX_CONNECTION_ATTEMPTS = 10

# Port to expect the docker environment to internally listen on.
_DOCKER_INTERNAL_GRPC_PORT = 10000

_DEFAULT_DOCKER_IMAGE_NAME = 'gcr.io/deepmind-environments/dm_fast_mapping:v1.1.1'

_FAST_MAPPING_TASK_OBSERVATIONS = ['RGB_INTERLEAVED', 'TEXT']

FAST_MAPPING_TASK_LEVEL_NAMES = frozenset((
    'architecture_comparison/fast_map_three_objs',
    'fast_slow/fast_map_three_objs_bed_tray',
    'fast_slow/fast_map_three_objs_bed_tray_putting_near',
    'fast_slow/fast_map_three_objs_bed_tray_putting_on',
    'fast_slow/fast_map_three_objs',
    'fast_slow/slow_learn_three_objs_bed_tray_lifting',
    'fast_slow/slow_learn_three_objs_bed_tray_putting_near',
    'fast_slow/slow_learn_three_objs_bed_tray_putting_on',
    'fast_slow/test_holdout_fast_map_three_objs_bed_tray_putting_on',
    'fast_slow/two_phase_slow_learn_three_objs_bed_tray_putting_near',
    'fast_slow/two_phase_slow_learn_three_objs_bed_tray_putting_on',
    'intrinsic_motivation/fast_map_three_objs_no_shaping_reward',
    'new_obj_generalization/fast_map_heldout_test_objs',
    'new_obj_generalization/fast_map_three_objs_global_five',
    'new_obj_generalization/fast_map_three_objs_global_ten',
    'new_obj_generalization/fast_map_three_objs_global_three',
    'new_obj_generalization/fast_map_three_objs_global_twenty',
    'num_generalization/fast_map_eight_objs',
    'num_generalization/fast_map_five_objs',
    'num_generalization/fast_map_three_objs',
    'with_distractors/eval_fast_map_two_episodes_three_objs_five_distractor',
    'with_distractors/eval_fast_map_three_episodes_three_objs_five_distractor',
    'with_distractors/eval_fast_map_four_episodes_three_objs_no_distractor',
    'with_distractors/eval_fast_map_four_episodes_three_objs_one_distractor',
    'with_distractors/eval_fast_map_three_objs_ten_distractor',
    'with_distractors/eval_fast_map_three_objs_twenty_distractor',
    'with_distractors/fast_map_three_objs_no_distractor',
    'with_distractors/fast_map_three_objs_one_distractor',
    'with_distractors/fast_map_three_objs_two_distractor',
))

_ConnectionDetails = collections.namedtuple('_ConnectionDetails',
                                            ['channel', 'connection', 'specs'])


class _FastMappingTasksEnv(dm_env_adaptor.DmEnvAdaptor):
  """An implementation of dm_env_rpc.DmEnvAdaptor for Fast Language Learning tasks."""

  def __init__(self, connection_details, requested_observations,
               num_action_repeats):
    super(_FastMappingTasksEnv,
          self).__init__(connection_details.connection,
                         connection_details.specs, requested_observations)
    self._channel = connection_details.channel
    self._num_action_repeats = num_action_repeats

  def close(self):
    super(_FastMappingTasksEnv, self).close()
    self._channel.close()

  def step(self, action):
    """Implementation of dm_env.step that supports repeated actions."""

    timestep = None
    discount = None
    reward = None
    for _ in range(self._num_action_repeats):
      next_timestep = super(_FastMappingTasksEnv, self).step(action)

      # Accumulate reward per timestep.
      if next_timestep.reward is not None:
        reward = (reward or 0.) + next_timestep.reward

      # Calculate the product for discount.
      if next_timestep.discount is not None:
        discount = discount if discount else []
        discount.append(next_timestep.discount)

      timestep = dm_env.TimeStep(next_timestep.step_type, reward,
                                 # Note: np.product(None) returns None.
                                 np.product(discount),
                                 next_timestep.observation)

      if timestep.last():
        return timestep

    return timestep


class _FastMappingTasksContainerEnv(_FastMappingTasksEnv):
  """An implementation of _FastMappingTasksEnv.

    Ensures that the provided Docker container is closed on exit.
  """

  def __init__(self, connection_details, requested_observations,
               num_action_repeats, container):
    super(_FastMappingTasksContainerEnv,
          self).__init__(connection_details, requested_observations,
                         num_action_repeats)
    self._container = container

  def close(self):
    super(_FastMappingTasksContainerEnv, self).close()
    try:
      self._container.kill()
    except docker.errors.NotFound:
      pass  # Ignore, container has already been closed.


class _FastMappingTasksProcessEnv(_FastMappingTasksEnv):
  """An implementation of _FastMappingTasksEnv.

    Ensure that the provided running process is closed on exit.
  """

  def __init__(self, connection_details, requested_observations,
               num_action_repeats, process):
    super(_FastMappingTasksProcessEnv,
          self).__init__(connection_details, requested_observations,
                         num_action_repeats)
    self._process = process

  def close(self):
    super(_FastMappingTasksProcessEnv, self).close()
    self._process.terminate()
    self._process.wait()


def _check_grpc_channel_ready(channel):
  """Helper function to check the gRPC channel is ready N times."""
  for _ in range(_MAX_CONNECTION_ATTEMPTS - 1):
    try:
      return grpc.channel_ready_future(channel).result(timeout=1)
    except grpc.FutureTimeoutError:
      pass
  return grpc.channel_ready_future(channel).result(timeout=1)


def _can_send_message(connection):
  """Returns if `connection` is healthy and able to process requests."""
  try:
    # This should return a response with an error unless the server isn't yet
    # receiving requests.
    connection.send(dm_env_rpc_pb2.StepRequest())
  except error.DmEnvRpcError:
    return True
  except grpc.RpcError:
    return False


def _create_channel_and_connection(port):
  """Returns a tuple of `(channel, connection)`."""
  for _ in range(_MAX_CONNECTION_ATTEMPTS):
    channel = grpc.secure_channel('localhost:{}'.format(port),
                                  grpc.local_channel_credentials())
    _check_grpc_channel_ready(channel)
    connection = dm_env_rpc_connection.Connection(channel)
    if _can_send_message(connection):
      break
    else:
      # A gRPC server running within Docker sometimes reports that the channel
      # is ready but transitively returns an error (status code 14) on first
      # use.  Giving the server some time to breath and retrying often fixes the
      # problem.
      connection.close()
      channel.close()
      time.sleep(1.0)

  return channel, connection


def _parse_exception_message(message):
  """Returns a human-readable version of a dm_env_rpc json error message."""
  try:
    match = re.match(r'^message\:\ \"(.*)\"$', message)
    json_data = codecs.decode(match.group(1), 'unicode-escape')
    parsed_json_data = json.loads(json_data)
    return ValueError(json.dumps(parsed_json_data, indent=4))
  except:  # pylint: disable=bare-except
    return message


def _wrap_send(send):
  """Wraps `send` in order to reformat exceptions."""
  try:
    return send()
  except ValueError as e:
    e.args = [_parse_exception_message(e.args[0])]
    raise


def _connect_to_environment(port, settings):
  """Helper function for connecting to a running dm_fast_mapping environment."""
  if settings.level_name not in FAST_MAPPING_TASK_LEVEL_NAMES:
    raise ValueError(
        'Level named "{}" is not a valid dm_fast_mapping level.'.format(
            settings.level_name))
  channel, connection = _create_channel_and_connection(port)
  original_send = connection.send
  connection.send = lambda request: _wrap_send(lambda: original_send(request))
  world_name = connection.send(
      dm_env_rpc_pb2.CreateWorldRequest(
          settings={
              'seed': tensor_utils.pack_tensor(settings.seed),
              'episodeId': tensor_utils.pack_tensor(0),
              'levelName': tensor_utils.pack_tensor(settings.level_name),
          })).world_name
  join_world_settings = {
      'width':
          tensor_utils.pack_tensor(settings.width),
      'height':
          tensor_utils.pack_tensor(settings.height),
      'EpisodeLengthSeconds':
          tensor_utils.pack_tensor(settings.episode_length_seconds),
      'ShowReachabilityHUD': tensor_utils.pack_tensor(False),
  }
  specs = connection.send(
      dm_env_rpc_pb2.JoinWorldRequest(
          world_name=world_name, settings=join_world_settings)).specs
  return _ConnectionDetails(channel=channel, connection=connection, specs=specs)


class EnvironmentSettings(typing.NamedTuple):
  """Collection of settings used to start a specific Fast Language Learning task.

    Required attributes:
      seed: Seed to initialize the environment's RNG.
      level_name: Name of the level to load.
    Optional attributes:
      width: Width (in pixels) of the desired RGB observation; defaults to 96.
      height: Height (in pixels) of the desired RGB observation; defaults to 72.
      episode_length_seconds: Maximum episode length (in seconds); defaults to
        120.
      num_action_repeats: Number of times to step the environment with the
        provided action in calls to `step()`.
  """
  seed: int
  level_name: str
  width: int = 96
  height: int = 72
  episode_length_seconds: float = 120.0
  num_action_repeats: int = 1


def _validate_environment_settings(settings):
  """Helper function to validate the provided environment settings."""
  if settings.episode_length_seconds <= 0.0:
    raise ValueError('episode_length_seconds must have a positive value.')
  if settings.num_action_repeats <= 0:
    raise ValueError('num_action_repeats must have a positive value.')
  if settings.width <= 0 or settings.height <= 0:
    raise ValueError('width and height must have a positive value.')
  if ('with_distractors/' in settings.level_name and
      settings.episode_length_seconds != 450.0):
    raise ValueError(
        'episode_length_seconds must be 450.0 for with_distractors/ levels.')


def load_from_disk(path, settings):
  """Load Fast Language Learning Tasks from disk.

  Args:
    path: Directory containing dm_fast_mapping environment.
    settings: EnvironmentSettings required to start the environment.

  Returns:
    An implementation of dm_env.Environment.

  Raises:
    RuntimeError: If unable to start environment process.
  """
  _validate_environment_settings(settings)

  executable_path = os.path.join(path, 'Linux64Player')
  libosmesa_path = os.path.join(path, 'external_libosmesa_llvmpipe.so')
  if not os.path.exists(executable_path) or not os.path.exists(libosmesa_path):
    raise RuntimeError(
        'Cannot find dm_fast_mapping executable or dependent files at path: {}'
        .format(path))

  port = portpicker.pick_unused_port()

  process_flags = [
      executable_path,
      # Unity command-line flags.
      '-logfile',
      '-batchmode',
      '-noaudio',
      # Other command-line flags.
      '--logtostderr',
      '--server_type=GRPC',
      '--uri_address=[::]:{}'.format(port),
  ]

  os.environ.update({
      'UNITY_RENDERER': 'software',
      'UNITY_OSMESA_PATH': libosmesa_path,
  })

  process = subprocess.Popen(
      process_flags, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
  if process.poll() is not None:
    raise RuntimeError('Failed to start dm_fast_mapping process correctly.')

  return _FastMappingTasksProcessEnv(
      _connect_to_environment(port, settings), _FAST_MAPPING_TASK_OBSERVATIONS,
      settings.num_action_repeats, process)


def load_from_docker(settings, name=None):
  """Load Fast Language Learning Tasks from docker container.

  Args:
    settings: EnvironmentSettings required to start the environment.
    name: Optional name of Docker image that contains the dm_fast_mapping
      environment. If left unset, uses the dm_fast_mapping default name.

  Returns:
    An implementation of dm_env.Environment
  """
  _validate_environment_settings(settings)

  name = name or _DEFAULT_DOCKER_IMAGE_NAME
  client = docker.from_env()

  port = portpicker.pick_unused_port()

  try:
    client.images.get(name)
  except docker.errors.ImageNotFound:
    logging.info('Downloading docker image "%s"...', name)
    client.images.pull(name)
    logging.info('Download finished.')

  container = client.containers.run(
      name,
      auto_remove=True,
      detach=True,
      ports={_DOCKER_INTERNAL_GRPC_PORT: port})

  return _FastMappingTasksContainerEnv(
      _connect_to_environment(port, settings), _FAST_MAPPING_TASK_OBSERVATIONS,
      settings.num_action_repeats, container)


examples/human_agent.py
# Copyright 2019 DeepMind Technologies Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Example human agent for interacting with DeepMind Fast Language Learning Tasks."""

from absl import app
from absl import flags
from absl import logging
import dm_fast_mapping
import numpy as np
import pygame

FLAGS = flags.FLAGS

flags.DEFINE_list(
    'screen_size', [640, 480],
    'Screen width/height in pixels. Scales the environment RGB observations to '
    'fit the screen size.')

flags.DEFINE_string(
    'docker_image_name', None,
    'Name of the Docker image that contains the Fast Language Learning Tasks. '
    'If None, uses the default dm_fast_mapping name')

flags.DEFINE_integer('seed', 123, 'Environment seed.')
flags.DEFINE_string('level_name', 'fast_slow/fast_map_three_objs',
                    'Name of task to run.')

_FRAMES_PER_SECOND = 30

_KEYS_TO_ACTION = {
    pygame.K_w: {'MOVE_BACK_FORWARD': 1},
    pygame.K_s: {'MOVE_BACK_FORWARD': -1},
    pygame.K_a: {'STRAFE_LEFT_RIGHT': -1},
    pygame.K_d: {'STRAFE_LEFT_RIGHT': 1},
    pygame.K_UP: {'LOOK_DOWN_UP': -1},
    pygame.K_DOWN: {'LOOK_DOWN_UP': 1},
    pygame.K_LEFT: {'LOOK_LEFT_RIGHT': -1},
    pygame.K_RIGHT: {'LOOK_LEFT_RIGHT': 1},
    pygame.K_SPACE: {'HAND_GRIP': 1},
}  # pyformat: disable
_NO_ACTION = {
    'MOVE_BACK_FORWARD': 0,
    'STRAFE_LEFT_RIGHT': 0,
    'LOOK_LEFT_RIGHT': 0,
    'LOOK_DOWN_UP': 0,
    'HAND_GRIP': 0,
}


def main(_):
  pygame.init()
  try:
    pygame.mixer.quit()
  except NotImplementedError:
    pass
  pygame.display.set_caption('Fast Language Learning Tasks Human Agent')

  if 'with_distractors' in FLAGS.level_name:  # for the tasks from the HTM paper
    episode_length_seconds = 450.0
  else:
    episode_length_seconds = 120.0
  env_settings = dm_fast_mapping.EnvironmentSettings(
      seed=FLAGS.seed, level_name=FLAGS.level_name,
      episode_length_seconds=episode_length_seconds)
  with dm_fast_mapping.load_from_docker(name=FLAGS.docker_image_name,
                                        settings=env_settings) as env:
    screen = pygame.display.set_mode(
        (int(FLAGS.screen_size[0]), int(FLAGS.screen_size[1])))

    rgb_spec = env.observation_spec()['RGB_INTERLEAVED']
    surface = pygame.Surface((rgb_spec.shape[1], rgb_spec.shape[0]))

    actions = _NO_ACTION
    score = 0
    clock = pygame.time.Clock()
    while True:
      # Do not close with CTRL-C as otherwise the docker container may be left
      # running on exit.
      for event in pygame.event.get():
        if event.type == pygame.QUIT:
          return
        elif event.type == pygame.KEYDOWN:
          if event.key == pygame.K_ESCAPE:
            return
          key_actions = _KEYS_TO_ACTION.get(event.key, {})
          for name, action in key_actions.items():
            actions[name] += action
        elif event.type == pygame.KEYUP:
          key_actions = _KEYS_TO_ACTION.get(event.key, {})
          for name, action in key_actions.items():
            actions[name] -= action

      timestep = env.step(actions)
      frame = np.swapaxes(timestep.observation['RGB_INTERLEAVED'], 0, 1)
      font = pygame.font.SysFont('Sans', 10)
      pygame.surfarray.blit_array(surface, frame)
      text = font.render(timestep.observation['TEXT'], True, (0, 0, 0))
      surface.blit(text, (0, 0))
      pygame.transform.smoothscale(surface, screen.get_size(), screen)

      pygame.display.update()

      if timestep.reward:
        score += timestep.reward
        logging.info('Total score: %1.1f, reward: %1.1f', score,
                     timestep.reward)
      clock.tick(_FRAMES_PER_SECOND)


if __name__ == '__main__':
  app.run(main)
